### DISCUSSION TITLE
Can someone explain what ggml_backend_sched_t do?

### DISCUSSION THREAD

**Discussion by Zijie-Tian** [ggml-org/llama.cpp#10182] [Q&A] [answered]
Recently, I’ve been studying ggml_backend_sched_t in ggml. I understand that `sched` enables compute with multi-backends. However, I’m quite confused about `ggml_backend_sched_split_graph`, `ggml_backend_sched_alloc_splits`, and `ggml_backend_sched_reserve`. Could someone help me clarify:

1. I understand that ggml_backend_sched_split_graph is likely dividing the computation graph based on buffer types, but in the logic of splitting, I have a hard time understanding the following code:
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1184-L1243
What is the purpose of this code?

2. Additionally, I’m still confused by the following code: 
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1357-L1400
I know it establishes a copy between different buffers, **but is this copy pre-allocated with a buffer, or is it copied at runtime?**

3. What is the difference between `ggml_backend_sched_reserve` and `ggml_backend_sched_alloc_graph`? I’m confused about the purpose of each.
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1727-L1755

4. I noticed that my code enters the following section and throws an error. When it enters this if branch, **I’m really curious about what this part is checking? how can i solve it by modify compute graph?**
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1524-L1534

5. **What is the purpose of `sched->is_alloc`?**
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1724
https://github.com/ggerganov/llama.cpp/blob/a9e8a9a0306a8093eef93b0022d9f45510490072/ggml/src/ggml-backend.cpp#L1752

**Comment by slaren** [comment, marked as answer]
1. I am not sure that I can add much more to what is already written in the comment. If you are confused why that is necessary, keep in mind that the scheduler does not find the most optimal way to split the graph algorithmically, because that would be far too expensive. Instead, it is a bunch of heuristics that work well in the common use cases. So the answer to why the split is done the way it is, it is because it solved some specific problem that was found during development.
2. The copies are performed during `ggml_backend_sched_compute_splits`. The reason there may be multiple copies of each tensor is because it improves efficiency when using pipeline parallelism.
3. reserve is only intended to be used to pre-allocate the compute buffers
4. To avoid this you need to reserve the graph with the largest possible tensors (eg. with the maximum context size in a transformer), and ensure that the following graphs are topologically identical and use tensors that are equal or smaller than the graph used during reserve. Generally this is not something that you need to worry about unless you are trying to implement pipeline parallelism. 
5. To keep track of whether the graph is allocated so that it can be allocated automatically in `ggml_backend_sched_graph_compute` if necessary

**Reply by Zijie-Tian** [reply]
Thank you for you answer! slaren!!! . 
> I am not sure that I can add much more to what is already written in the comment. If you are confused why that is necessary, keep in mind that the scheduler does not find the most optimal way to split the graph algorithmically, because that would be far too expensive. Instead, it is a bunch of heuristics that work well in the common use cases. So the answer to why the split is done the way it is, it is because it solved some specific problem that was found during development.

 As I understand it, currently, ggml_backend_sched_compute_splits is only used to identify the positions where additional copies are needed and to promote operators to a higher-priority backend (e.g., the GPU) whenever possible. Is it right?

> To avoid this you need to reserve the graph with the largest possible tensors (eg. with the maximum context size in a transformer), and ensure that the following graphs are topologically identical and use tensors that are equal or smaller than the graph used during reserve. Generally this is not something that you need to worry about unless you are trying to implement pipeline parallelism.

Additionally, I would still appreciate a detailed explanation of the cause of this error. **Could you perhaps give a very simple example?**

**Reply by Zijie-Tian** [reply]
> The copies are performed during ggml_backend_sched_compute_splits. The reason there may be multiple copies of each tensor is because it improves efficiency when using pipeline parallelism.

Additionally, I'd like to ask if the intermediate computation results (model activations) will cause multiple memory copy (for example, Qcur in CPU and GPU?) during inference runtime?

**Reply by slaren** [reply]
> As I understand it, currently, ggml_backend_sched_compute_splits is only used to identify the positions where additional copies are needed and to promote operators to a higher-priority backend (e.g., the GPU) whenever possible. Is it right?

`ggml_backend_sched_split_graph` does that. `ggml_backend_sched_compute_splits` does the computation, copying data between the graph splits as needed.

> Additionally, I would still appreciate a detailed explanation of the cause of this error. Could you perhaps give a very simple example?

This is not an error, it's just a message for debugging issues. It happens when evaluating a graph that either has a different topology than the previous graph (ie. the graph nodes are different), or when the topology is the same, but some tensors are larger. This means that the memory addresses assigned to each tensor need to be recalculated, and this forces a synchronization. Generally this is not something that you need to worry about.

> Additionally, I'd like to ask if the intermediate computation results (model activations) will cause multiple memory copy (for example, Qcur in CPU and GPU?) during inference runtime?

If you set the environment variable `GGML_SCHED_DEBUG` to 1 or 2, you will get a detailed report of how the graph was split. At the start of every split, you will see a list of inputs: these are tensors that are being copied to a different backend.

**Reply by Zijie-Tian** [reply]
Thank you!!! Slaren. I greatly appreciate your work.

**Comment by Zijie-Tian** [comment]
slaren, I found that in the `ggml_backend_sched_split_graph` function, there are extensive graph splitting operations internally, depending on the backend, which results in multiple splits. **However, these splits may lead to unnecessary memory copies in runtime**. I'm wondering if there could be cases where different backends don't actually require splitting. For example, Mac or Jetson's Unified Memory might not need these splittings.
